src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+package book.store.service.impl;
+
+import book.store.dto.cartitem.CartItemRequestDto;
+import book.store.dto.cartitem.UpdateCartItemDto;
+import book.store.dto.shoppingcart.ShoppingCartDto;
+import book.store.exception.EntityNotFoundException;
+import book.store.mapper.CartItemMapper;
+import book.store.mapper.ShoppingCartMapper;
+import book.store.model.CartItem;
+import book.store.model.ShoppingCart;
+import book.store.repository.cartitem.CartItemRepository;
+import book.store.repository.shoppingcart.ShoppingCartRepository;
+import book.store.service.ShoppingCartService;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+
+    @Transactional
+    @Override
+    public ShoppingCartDto addBookInShoppingCart(Long userId, CartItemRequestDto requestDto) {
+        ShoppingCart userShoppingCart = getShoppingCartByUserId(userId);
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+
+        Optional<CartItem> cartItemByBookId = userShoppingCart.getCartItems().stream()
+                .filter(cartItemBook -> cartItemBook.getBook().getId()
+                .equals(requestDto.getBookId()))
+                .findFirst();
+
+        if (cartItemByBookId.isEmpty()) {
+            cartItemRepository.save(cartItem);
+            userShoppingCart.getCartItems().add(cartItem);
+            return shoppingCartMapper.toDto(userShoppingCart);
+        }
+
+        CartItem cartItemWithoutNewQuantity = cartItemByBookId.orElseThrow(() ->
+            new EntityNotFoundException("Can`t find cart item by id: " + requestDto.getBookId()));
+
+        int oldQuantity = cartItemWithoutNewQuantity.getQuantity();
+        cartItemRepository.delete(cartItemWithoutNewQuantity);
+        cartItem.setQuantity(cartItem.getQuantity() + oldQuantity);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(userShoppingCart);
```

Comment:
it's kind of hard to read this method, please, extract different logic to local methods